validator/src/main/java/org/ethereum/beacon/validator/MultiValidatorService.java

```diff
@@ -0,0 +1,375 @@
+
+  private static final Logger logger = LogManager.getLogger(MultiValidatorService.class);
+
+  private final Schedulers schedulers;
+
+  private final DirectProcessor<BeaconBlock> blocksSink = DirectProcessor.create();
+  private final Publisher<BeaconBlock> blocksStream;
+
+  private final DirectProcessor<Attestation> attestationsSink = DirectProcessor.create();
+  private final Publisher<Attestation> attestationsStream;
+
+  /** Proposer logic. */
+  private BeaconChainProposer proposer;
+  /** Attester logic. */
+  private BeaconChainAttester attester;
+  /** The spec. */
+  private SpecHelpers specHelpers;
+
+  private Publisher<ObservableBeaconState> stateStream;
+
+  /** Credentials of yet not initialized validators. */
+  private Map<BLSPubkey, BLS381Credentials> uninitialized;
+  /** Credentials of already initialized validators. */
+  private Map<ValidatorIndex, BLS381Credentials> initialized = new ConcurrentHashMap<>();
+  /** Latest slot that has been processed. */
+  private SlotNumber lastProcessedSlot = SlotNumber.ZERO;
+  /** The most recent beacon state came from the outside. */
+  private ObservableBeaconState recentState;
+
+  /** Validator task executor. */
+  private final Scheduler executor;
+
+  private final LatestExecutor<BeaconState> initExecutor;
+
+  public MultiValidatorService(
+      List<BLS381Credentials> blsCredentials,
+      BeaconChainProposer proposer,
+      BeaconChainAttester attester,
+      SpecHelpers specHelpers,
+      Publisher<ObservableBeaconState> stateStream,
+      Schedulers schedulers) {
+    this.uninitialized =
+        blsCredentials.stream()
+            .collect(Collectors.toMap(BLS381Credentials::getPubkey, Function.identity()));
+    this.proposer = proposer;
+    this.attester = attester;
+    this.specHelpers = specHelpers;
+    this.stateStream = stateStream;
+    this.schedulers = schedulers;
+
+    blocksStream =
+        Flux.from(blocksSink)
+            .publishOn(this.schedulers.reactorEvents())
+            .onBackpressureError()
+            .name("BeaconChainValidator.block");
+
+    attestationsStream =
+        Flux.from(attestationsSink)
+            .publishOn(this.schedulers.reactorEvents())
+            .onBackpressureError()
+            .name("BeaconChainValidator.attestation");
+
+    executor = this.schedulers.newSingleThreadDaemon("validator-service");
+    initExecutor = new LatestExecutor<>(schedulers.cpuHeavy(), this::initFromLatestBeaconState);
+  }
+
+  @Override
+  public void start() {
+    subscribeToStateUpdates(this::onNewState);
+  }
+
+  @Override
+  public void stop() {}
+
+  /**
+   * Passes validators from {@link #uninitialized} map through initialization process.
+   *
+   * @param state a state object to seek for validators in.
+   */
+  private void initFromLatestBeaconState(BeaconState state) {
+    Map<ValidatorIndex, BLS381Credentials> intoCommittees = new HashMap<>();
+    for (ValidatorIndex i : state.getValidatorRegistry().size()) {
```

Comment:
Potential place for optimization: it would be too heavy to iterate over all validators each new `BeaconState`.